up-l1/mqtt.adoc

```diff
@@ -129,6 +132,23 @@ UUri {
 
 Note: The kind of message should be determined using the Type UAttribute in the User Properites of the message.
 
+If a UUri is provided with only a UAuthority we should subscribe to all topics under the UAuthority:
+
+[source]
+----
+UUri {
+  authority: UAuthority {
+    name: "Vehicle_1",
+    number: 192.168.0.0,
+  },
+  entity: empty,
+  resource: empty,
+}
+----
+
+* If in cloud, uses wildcard to subscribe to all messages from the UAuthority: `d2c/192.168.0.0/+/+/+`
+* If on device, uses wildcard to subscribe to all messages from the UAuthority: `c2d/192.168.0.0/+/+/+`
```

Comment:
Is "in cloud" vs "on device" determined by when we construct a `UPClientMQTT5`, passing in a parameter to dictate that?

Reply:
What if we want to get messages from the cloud at `uauthority_1` and messages from other devices locally within the vehicle at `uauthority_2`?

Want to make sure we're sort of not limiting ourselves into a certain topology here prematurely :slightly_smiling_face:

Tagging in @stevenhartley for sanity check

Reply:
To your first question, yes my thought was the UPClient would have some initialization value to determine if its running on the device and connected to a remote mqtt, running on the device and connected to a local mqtt or running in the cloud.

Reply:
However, i do see your point that multiple hops may break this system down

Reply:
Okay one concern I have in general is how EntityA on a device with no direct connection to the cloud signals to EntityB (with direct connection to the cloud) that it wants to receive information from a different uAuthority. EntityB will have 2 upClients (one connected to the cloud mqtt and one connected to the remote mqtt).

Reply:
A simple register listener /uAuthority_2/+/+/+ doesnt work for EntityA since EntityB doesnt know that EntityA is listening on that topic. EntityB would need to get some form of notification that EntityA is listening on that remote topic for the local mqtt broker so that it can listen to the remote topic as well (on the cloud mqtt broker) to forward along the messages to the local mqtt broker

Reply:
> Okay one concern I have in general is how EntityA on a device with no direct connection to the cloud signals to EntityB (with direct connection to the cloud) that it wants to receive information from a different uAuthority. EntityB will have 2 upClients (one connected to the cloud mqtt and one connected to the remote mqtt).

No the D2D communication is through the streamer vs uE-2-uE communication is through a MQTT5 broker. So a uE only uses one up-client library talking to a local broker.  The streamer yes will have 2 different up-client (uTransport) instances, one for talking to other devices (streamers) and one to talk to the local broker.

Reply:
@PLeVasseur @stevenhartley @lynn2009  What are your thoughts on this approach?:

location_identifier:
- upr: Remote destination
- upl: Local destination

topic uri mapping:

  `{location_identifier}/{UUri.UAuthority.number (remote only)}/{UUri.UEntity.id}/{UUri.UEntity.version_major}/{UUri.UResource.id}`

Publish uses source UUri and all other message types use sink UUri when sending a message.

subscribing (registerListener):

  If subscribing to a different UAuthority, will subscribe to all Publish message types on the UUri:

    `upr/{UUri.UAuthority.number}/+/+/+` where `UAttribute.type == Publish`

  if subscribing to the own authority, will subscribe to all messages except Publish message type:

    `upr/{UUri.UAuthority.number}/+/+/+` where `UAttribute.type != Publish`

This should avoid the scenario where an authority is publishing messages to the cloud and trying to receive messages at the same topic since source and sink may match.

Alternatively, the Listener could sort messages based on whether the `source` and `topic` match at the uAuthority level.

Reply:
Hey @devkelley -- sorry, I missed this.

We ended up rolling back our thoughts about separating `upl/` vs `upr/` over in the Zenoh case and Zenoh always does a "global" send onto `up/`. Is the concept of remote vs local at the key/header level still needed here for MQTT? (again, don't know MQTT at all yet).

-------

Speaking from a uStreamer perspective:

> if subscribing to the own authority, will subscribe to all messages except Publish message type:
> 
> ```
> `upr/{UUri.UAuthority.number}/+/+/+` where `UAttribute.type != Publish`
> ```

This matches what I'd expect. In fact, because I assume that this portion:
> where `UAttribute.type != Publish`

would not be filtered for, I currently [have an explicit check to throw away any Publish messages](https://github.com/eclipse-uprotocol/up-streamer-rust/pull/7/files#r1548303397) in my initial PR for the `up-streamer-rust`. :slightly_smiling_face: 

> If subscribing to a different UAuthority, will subscribe to all Publish message types on the UUri:
> 
> ```
> `upr/{UUri.UAuthority.number}/+/+/+` where `UAttribute.type == Publish`
> ```

Again, another feature that's unexpected, but welcome :slightly_smiling_face: I was assuming I'd have to do a similar check to throw away any non-Publish messages.

Reply:
> Okay one concern I have in general is how EntityA on a device with no direct connection to the cloud signals to EntityB (with direct connection to the cloud) that it wants to receive information from a different uAuthority. EntityB will have 2 upClients (one connected to the cloud mqtt and one connected to the remote mqtt).

Could you draw me a simple picture / diagram for this case and help me understand the challenge for MQTT in this case? Perhaps this comes up because MQTT is brokered, unlike Zenoh? Again total MQTT newb.

Reply:
![image](https://github.com/eclipse-uprotocol/up-spec/assets/105753233/a6a4a660-2a2a-4b7a-9b06-124765aea8a9)
So maybe the context I was missing is that EntityB in this scenario would be a uStreamer. But the issue I was thinking about is that the EntityA will subscribe to `up/Authority2/+/+/+` on the local mqtt broker (Orange). How does EntityB get notified that EntityA has a listener for `up/Authority2/+/+/+` since there is no notification that a listener was setup by EntityA?

However, @stevenhartley mentioned that for device to device communication a streamer would be used, so maybe this doesnt apply?

Reply:
I'm going to speak to this as if we're talking about Publish messages only for the moment.

I think that only a uStreamer would be subscribing on the UAuthority alone.

Therefore, if the uE `EntityA` wanted to register for a particular Publish message called `up/Authority2/EntityFoo/ResourceBar`, that'd be coordinated through uSubscription. `EntityA` would call [`Subscribe()`](https://github.com/eclipse-uprotocol/up-spec/blob/36c4a2ded83b5d69fdabca3d7122d70ffa7556b7/up-core-api/uprotocol/core/usubscription/v3/usubscription.proto#L62) on the client uSubscription object, which would then coordinate with the uSubscription service on the device for `Authority1`.

The uSubscription service in the background would then check that this topic is being piped over to us from the cloud (or any other connected device) and if not, make it so.

In addition, the uStreamer would also get signaled that there was a change in subscribers via [`RegisterForNotifications()`](https://github.com/eclipse-uprotocol/up-spec/blob/36c4a2ded83b5d69fdabca3d7122d70ffa7556b7/up-core-api/uprotocol/core/usubscription/v3/usubscription.proto#L93) and perform the routing appropriately so that any `up/Authority2/EntityFoo/ResourceBar` get picked up from the Cloud MQTT and put onto the Local MQTT.

------

Some questions to help me make sure I'm understanding correctly:
1. Are the Local MQTT and Cloud MQTT completely different implementations of `UTransport`? Same?
2. Are the Local MQTT and Cloud MQTT unable to "see" each other due to the brokered nature of MQTT?

Reply:
1. My intention is that the Local and Cloud clients would be the same implementation of UTransport, just connecting to the different brokers. If this feels too difficult we could change to different implementations.
2. Correct, in this scenario they would not have knowledge of each other and the 'entity' that is using the upclients would be responsible for handling messages between the two clients.

Reply:
Thanks for answers to 1 & 2. Makes sense to me.

I think what I wrote in my above response holds then. In this case `EntityB` would be a uStreamer to bridge the Local MQTT and Cloud MQTT.

That uStreamer would have 2 x instances of `up-client-mqtt5-rust`, one used with the Local MQTT broker and one for the Remote MQTT broker.

Upon getting a signal via uSubscription that _some entity_ on `UAuthority1` wanted to subscribe to `up/Authority2/EntityFoo/ResourceBar` which is on `UAuthority2`, then the uStreamer would subscribe to `up/Authority2/EntityFoo/ResourceBar` from the Cloud MQTT via the `up-client-mqtt5-rust` `UTransport` for the Cloud and publish that same message, but onto the `up-client-mqtt5-rust` `UTransport` for the Local MQTT broker.

Reply:
Okay, I think this was the context that I was missing. Essentially there is some higher level component that will handle signaling around the messages and the `up-client-mqtt5-rust` client will only be responsible for handling a single communication hop. Given this context I think we can drop the `remote` and `local` tagging.

Reply:
@PLeVasseur and just to confirm for zenoh: are there still two kinds of topics depending on local vs remote destination? ie.

up/{Authority}/{Entity}/{Resource}/  for remote

and

up/{Entity}/{Resource}/ for local?

Reply:
@devkelley -- we decided on the Zenoh Key Expressions always being global. So when sending any Publish, Request, Response, or Notification they are `up/UAuthority/UEntity/UResource` in short form.

I don't think it's reflected into the spec yet, though.

@evshary -- can you sanity check me?

Reply:
Ah got it, and if the uAuthority is not present when passed to a function it is added?

Reply:
Okay I updated the uuri mapping to what my understanding currently is, let me know if it doesnt match your understanding :)

Reply:
@PLeVasseur You're right. The spec hasn't been updated yet. I'm working on that.

Reply:
Initial version here
https://github.com/eclipse-uprotocol/up-spec/pull/103